test/app-tap/module_api.c

```diff
@@ -1711,6 +1715,83 @@ test_key_def_merge(struct lua_State *L)
 	key_def_check_merge(a_27, lengthof(a_27), b_27, lengthof(b_27),
 			    exp_27, lengthof(exp_27));
 
+	/*
+	 * Case 28: sort_order = 'asc' + sort_order = 'desc'.
+	 *
+	 * Interpretation: the same as for the case 3.
+	 */
+	box_key_part_def_t a_28[] = {
+		{{3, 0, "unsigned", NULL, NULL}},
+		{{1, 0, "unsigned", NULL, NULL}}, /* clash */
+	};
+	box_key_part_def_t b_28[] = {
+		{{1, 4, "unsigned", NULL, NULL}}, /* clash */
+		{{2, 0, "unsigned", NULL, NULL}},
+	};
+	box_key_part_def_t exp_28[] = {
+		{{3, 0, "unsigned", NULL, NULL}},
+		{{1, 0, "unsigned", NULL, NULL}}, /* coalesced */
+		{{2, 0, "unsigned", NULL, NULL}},
+	};
+	key_def_check_merge(a_28, lengthof(a_28), b_28, lengthof(b_28),
+			    exp_28, lengthof(exp_28));
```

Comment:
There is a comment (written by me) with attempt to formalize invariants of the `box_key_def_merge()` function. Quoted below.

> For any given kd1 and kd2, kd3 = merge(kd1, kd2) should impose the same order of tuples as if they would be ordered by kd1, but all kd1-equal tuples would be ordered by kd2.

(It is in the beginning of the `test_key_def_merge()` function.)

It seems, this comment is incorrect now. I think I know how to fix it:

> <...> but all kd1-equal tuples **that are not kd2-equal** would be ordered by kd2.

If you have a better description of merge's invariants or better know reasons behind them, I would appreciate if you enhance my description. At now, it is based on observations. It was written without good understanding of internals (such as comparators, validators and key extractors), but, as I see, you actively work on this code and likely know more than me.

----

Your interpretation of the test cases is a reference to the case 3, which is technically applicable to cases 28-29, but IMHO it doesn't really describes why the invariants are hold.

I think that if you describe why the fields are coalesced in these test cases in your words, it would be much better.

Reply:
Updated the test comment and slightly expanded your description, but it seems to me it's still correct: even if the equal-by-kd1 tuples are also equal-by-kd2, the tuples inside groups of kd-1-equal tuples are ordered by the kd2, even though the order is not changed.